llvm/docs/LangRef.rst

```diff
@@ -8021,6 +8021,43 @@ it will contain a list of ids, including the ids of the callsites in the
 full inline sequence, in order from the leaf-most call's id to the outermost
 inlined call.
 
+
+'``noalias.addrspace``' Metadata
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``noalias.addrspace`` metadata is used to identify memory
+operations which cannot access a range of address spaces. It is
+attached to memory instructions, including :ref:`atomicrmw
+<i_atomicrmw>`, :ref:`cmpxchg <i_cmpxchg>`, and :ref:`call <i_call>`
+instructions.
+
+This follows the same form as :ref:`range metadata <range-metadata>`,
+except the field entries must be of type `i32`. The interpretation is
+the same numeric address spaces as applied to IR values.
+
+Example:
+
+.. code-block:: llvm
+
+    ; %ptr cannot point to an object allocated in addrspace(5)
+    %rmw.valid = atomicrmw and ptr %ptr, i64 %value seq_cst, !noalias.addrspace !0
+
+    ; Undefined behavior. The underlying object is allocated in one of the listed
+    ; address spaces.
+    %alloca = alloca i64, addrspace(5)
+    %alloca.cast = addrspacecast ptr addrspace(5) %alloca to ptr
+    %rmw.ub = atomicrmw and ptr %alloca.cast, i64 %value seq_cst, !noalias.addrspace !0
+
+    !0 = !{i32 5, i32 6}
```

Comment:
Why does this say both 5 and 6 ?

Reply:
No, ranges are specified as an open range on the end. 5, 6 only excludes 5. { 5, 7 } would exclude 5 and 6

Reply:
It would probably make sense to add a comment here.

Reply:
Agreed, this is not obvious, I am only now realising I messed that up when I asked my question to clarify another aspect of this, and it went unnoticed.

Are ranges of address spaces, in general, a meaningful thing? I would not have assumed that current address space numbering is intended to be meaningful, there is no presumption that two address spaces being consecutive means they are any more or less likely to make sense to combine than two non-consecutive address spaces. And would excluding address spaces 5 and 7, but not 6, then require using two `!noalias.addrspace` annotations on the same instruction, or is there a better way? Whatever the answer is, could you add an example of that?

Reply:
This is only a way to add flexibility in the encoding for future use. There's no inherent meaning of consecutive address spaces.

In the current AMDGPU/PTX use cases, there's only one stack address space to consider. If in the future we wanted to reserve some range of address space values for software uses that may be allocated as stack, you could represent this without listing every possible value. Plus reusing all of the range parse and check code is nicer than needing to brute force search through a list of entries

Reply:
Thanks for the clarification.
FWIW I think that encoding this as a range instead of a list is fine.

Reply:
I can think of use cases where the set of address spaces is non-contiguous and the range base representation can be clunky as each range will be of size 1 and any coalescing is coincidental. Would it be possible to support both list and ranges, for example:

```
!0 = !{!"range" i32 5, i32 6} // a list or ranges follows
!0 = !{!"list" i32 5} or a list of address spaces follows
```

Reply:
> !0 = !{!"range" i32 5, i32 6} // a list of ranges follows

Though this is syntactic sugar that can some later if desired.